File: 1696-jump-game-vi/1696-jump-game-vi.py
class Solution:
    def maxResult(self, nums: List[int], k: int) -> int:
        
        ans = 0
        negs = []
        temp = [0]
        for i in range(len(nums)):
            if nums[i] >= 0 or i == 0 or i == len(nums) - 1:
                ans += nums[i]
                if len(temp) > k:
                    negs.append(temp)
                    temp = [0]
                else:
                    temp = [0]
            else:
                temp.append(nums[i])
        
        mem = dict()
        def res(pos = 0):
            if pos in mem:
                return mem[pos]
            elif pos > len(l):
                return -math.inf
            elif pos == len(l):
                return 0
            else:
                mxPos = None
                mx = -math.inf
                if pos + k >= len(l):
                    mem[pos] = l[pos]
                    return l[pos]
                for i in range(1, min(k+1, len(l) - pos + 1)):
                    if mx <= l[pos + i]:
                        mx = l[pos + i]
                        mxPos = i
                ret = -math.inf
                for i in range(mxPos, min(k+1, len(l) - pos + 1)):
                    ret = max(ret, res(pos+i))
                mem[pos] = ret + l[pos]
                return ret + l[pos]
        
        for l in negs:
            mem = dict()
            ans += res()
        
        return ans
                    

File: 1696-jump-game-vi/1696-jump-game-vi.java
class Solution {
    
    public int maxResult(int[] nums, int k) {
        
        PriorityQueue<Integer[]> q = new PriorityQueue<>(new Comparator<Integer[]>(){
            @Override
            public int compare(Integer[] o1, Integer[] o2) {
                if (o1[0] == o2[0]) {
                    return -Integer.compare(o1[1], o2[1]);
                }
                else {
                    return -Integer.compare(o1[0], o2[0]);
                }
            }
        });
        
        q.add(new Integer[] {nums[0], 0});
        
        for (int i = 1; i < nums.length; i++) {
            while (i - q.peek()[1] > k) {
                q.poll();
            }
            if (i == nums.length - 1) {
                return q.peek()[0] + nums[i];
            }
            else {
                q.add(new Integer[] {q.peek()[0] + nums[i], i});   
            }
        }
        return nums[0];
    }
    
    
    
}